### ISSUE TITLE
Intermediate cert and root CA must come via same source for verify to succeed

### ISSUE THREAD

**Issue by tuminoid** [sigstore/cosign#3976]
**Description**

Intermediate cert and root CA must come in via same source for verify to succeed.

**Expected behavior**

As long as chain of trust is complete, signing cert + intermediate cert + root CA should be able to be attached in registry or supplied to cosign verify via command line arguments. Specifically, it should be possible to have signing cert + intermediate cert attached to image in registry, while supplying only root CA for verification.

**Actual behavior**

1. intermediate cert + root CA are both attached as certchain using `cosign attach --cert-chain ...`, then verifiation versus root CA only works (but attached root cert is duplicated in two places)
2. if intermediate only is attached as cert chain, then verification with root CA only FAILS. In this case intermediate has to be supplied with root CA to verify (intermediate is duplicated in two places).

**Reproduction**

f I remove the root CA from the cert chain file [here](https://github.com/Nordix/metal3-clusterapi-docs/blob/226bedb94fa36220dd1176aaf2bd277a0842c639/security/container-signing-poc/cosign/gencrt.sh#L140), then cosign verify [here](https://github.com/Nordix/metal3-clusterapi-docs/blob/226bedb94fa36220dd1176aaf2bd277a0842c639/security/container-signing-poc/cosign/Makefile#L65) fails.

Relevant parts:

```text
# fails
cat "${OUTPUT_DIR}"/sub-ca.crt  > "${OUTPUT_DIR}"/certificate_chain.pem
# works
cat "${OUTPUT_DIR}"/sub-ca.crt "${OUTPUT_DIR}"/ca.crt > "${OUTPUT_DIR}"/certificate_chain.pem

cosign attach signature \
	--payload examples/payload.json \
	--signature examples/payloadbase64.sig \
	--certificate examples/leaf.crt \
	--certificate-chain examples/certificate_chain.pem \
	127.0.0.1:5003/alpine@sha256:33735bd63cf84d7e388d9f6d297d348c523c044410f553bd878c6d7829612735

cosign verify \
	--ca-roots=examples/ca.crt \
	--certificate-identity-regexp '.*' \
	--certificate-oidc-issuer-regexp '.*' \
	--private-infrastructure \
	--insecure-ignore-sct \
127.0.0.1:5003/alpine@sha256:33735bd63cf84d7e388d9f6d297d348c523c044410f553bd878c6d7829612735

Error: no matching signatures: cert verification failed: x509: certificate signed by unknown authority. Check your TUF root (see cosign initialize) or set a custom root with env var SIGSTORE_ROOT_FILE
main.go:69: error during command execution: no matching signatures: cert verification failed: x509: certificate signed by unknown authority. Check your TUF root (see cosign initialize) or set a custom root with env var SIGSTORE_ROOT_FILE
```

**Version**

```text
❯ cosign version
  ______   ______        _______. __    _______ .__   __.
 /      | /  __  \      /       ||  |  /  _____||  \ |  |
|  ,----'|  |  |  |    |   (----`|  | |  |  __  |   \|  |
|  |     |  |  |  |     \   \    |  | |  | |_ | |  . `  |
|  `----.|  `--'  | .----)   |   |  | |  |__| | |  |\   |
 \______| \______/  |_______/    |__|  \______| |__| \__|
cosign: A tool for Container Signing, Verification and Storage in an OCI registry.

GitVersion:    v2.4.1
GitCommit:     unknown
GitTreeState:  unknown
BuildDate:     unknown
GoVersion:     go1.23.1
Compiler:      gc
Platform:      linux/amd64
```

**References**

1. [Slack discussion](https://sigstore.slack.com/archives/C01PZKDL4DP/p1733150465358289)

**Comment by Eljees** [comment]
While looking at this I found something adjacent that is easy to state precisely, though I am not sure it is the same path you hit.

`ValidateAndUnpackCertWithChain` (`pkg/cosign/verify.go`) does not just add the supplied chain — it **replaces** the caller's roots with the chain's last element:

```go
rootPool := x509.NewCertPool()
rootPool.AddCert(chain[len(chain)-1])
co.RootCerts = rootPool          // whatever the caller put in RootCerts is dropped
```

Consequence, measured on `11926fa` with a throwaway test built from the existing helpers (`test.GenerateRootCa` / `GenerateSubordinateCa` / `GenerateLeafCert`):

* leaf signed by an intermediate, only the intermediate passed as the chain, and the **correct** root in `co.RootCerts` — verification succeeds, but because the intermediate was promoted to root, not because the root matched;
* same setup with a **completely unrelated** root in `co.RootCerts` — verification **also succeeds**. It should not. A test asserting an error here fails on a clean tree today.

So in this function the supplied roots are inert. That matches the shape of what you describe — "intermediate has to be supplied together with the root CA" — because the chain has to carry its own anchor.

The caveat, and the reason I am not opening a PR: this function is reached from `cmd/cosign/cli/verify/common.go` for the `--certificate-chain` flag, whereas your scenario attaches the chain in the registry, which comes in via `sig.Chain()` elsewhere. Those may be two different paths, and this is verification code — I would rather have a maintainer say which one is in scope than guess.

If it is useful, the change I tested was: keep `co.RootCerts` when the caller provided any, and treat the whole supplied chain as intermediates; fall back to today's behaviour (last element becomes the root) only when no roots were given. With that, the two tests above pass and the existing four `TestValidateAndUnpackCertWithChain*` tests plus the rest of `pkg/cosign` stay green. Happy to send it if that is the direction you want.

AI-assisted (LLM used for drafting); the runs above are mine.

**Comment by tuminoid** [comment]
Thanks for checking. I see in your PR that was closed the whole code path was removed.  I haven't been able to dedicate much time to cosign lately, so IIRC bundles are the only way cosign v4 is planned to work and the deprecated legacy paths are dropped?

**Comment by Hayden-IO** [comment]
Yes, that's what's planned. Would you be able to verify if this is still an issue in Cosign v3? The defaults will use the newer codepath. We can definitely fix this issue if it is still an issue, it just would need to be fixed in a different codepath.

**Comment by tuminoid** [comment]
It was bit of an adventure as I did not first notice it required v3.1.3 with sigstore-go v1.2.2 exactly, and some other trickery with new bundles considering I'm approaching this from BYOK POV.

TL;DR: with 3.1.3 it works as expected. 

| Path | Chain source | Anchor | Result |
|---|---|---|---|
| New bundle + --allow-certificate-chain | intermediate in bundle | root only | PASS |
| New bundle, flag omitted | intermediate in bundle | root only | no signatures found (silent legacy fallback) |
| Legacy attach | intermediate attached | root only | exit 12, unknown authority |
| Legacy attach | intermediate attached | root + --ca-intermediates | PASS |


As a follow-up question (again, not up to speed with new bundles at all), I could make it work only via "cosign import-key-pair", ie. giving up my keys to cosign. How is BYOK (external signers) supposed to work with v4/bundles, when attach is gone?

**Comment by Hayden-IO** [comment]
Thanks for checking on this, I'm glad to hear this is working as intended.

BYOK will work as well, the different bundle format and signing codepath didn't remove any features. You'll see in the bundle a "key hint" along with the signature. Here's an example, and if you wanted entirely offline signing, you could do so as well where you wouldn't fetch a timestamp or log the signature.

```
{
  "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
  "verificationMaterial": {
    "publicKey": {
      "hint": "eriUMy0szRptoBV+VoXkvQUHxs8ukSmAN9o9J1tSPzY="
    },
    "messageSignature": {
      "messageDigest": {
        "algorithm": "SHA2_256",
        "digest": "t63ISUBAFyzdBX/BzHMSb6sK4kIbQBTL302guoF20Bc="
      },
      "signature": "MEQCIAh3LvwwzqZaaL3F6cwZa5vqFQW0WHaFHwqL0WTt4Ix8AiA4SZfxnoVYDOxtEtgFDILExjvak/yNwkoIeITUq1DboQ=="
    },
    "tlogEntries": [
      {
        "logIndex": "2453706389",
        "logId": {
          "keyId": "wNI9atQGlz+VWfO6LRygH4QUfY/8W4RFwiT5i5WRgB0="
        },
        "kindVersion": {
          "kind": "hashedrekord",
          "version": "0.0.1"
        },
        "integratedTime": "1786641892",
        "inclusionPromise": {
          "signedEntryTimestamp": "MEUCIGY2Y37l5cUWJTIainBlsuFjfToTr7bwHDd79YD1o7hlAiEArIe3zesPQ7t7cnEZIxFIxLaTSHTHZzQ9EFCx0xGyKIs="
        },
        "inclusionProof": {
          "logIndex": "2331802127",
          "rootHash": "FoD8lX0hTW8FS7Qv5jtr/GnckUxCUSFL9JadwU76yPM=",
          "treeSize": "2331802244",
          "hashes": [
            "bIJLIdIdEgMTmYiGLp5BQTJdYOf0vg012U4FIoOYf9E=",
            "lfASFJ9q0fcPe6r22UuCyCnm2XxkEXXBA6xmQNk2Xmk=",
            "Jd8ByxGydXJ36/ShMBSzce+jdvRMX4oQtUDADG5G1Pw=",
            "BpQ+sR54IQdiWrPbV8FShnWw6gnrT0NBehhMSUi7aao=",
            "FQEn0sYehC9Jyd62YiuFAcZriRiMUL0UlryRvfBV708=",
            "aG7RSZNKV59pWCp3Hyspj0SfEeuF8ijf2wQbMyfO9fE=",
            "f450dCU0ptx4X8ADWbNfb6ETlJvmhv6X719wl70eQxw=",
            "zsDUAE1kJg8GGu4kwpIyvL9KK8xAnPt3R0cC8YV3gI4=",
            "D1bbdgPCHDQ8RPd5WpEsCiT2Q0DfVd3ZxhIJCblu0jU=",
            "gTT/NP592rDmkQgp49HUpBFgiYuWMohYIAmqc2AOpAM=",
            "OrFEEOUFSHYgMP4x3dQ55CR81JwPusy08V/8EL6he7Q=",
            "VAXTjEvkDlafwrSdzbB9EAdnyngfTL/1zJIl+oViS3E=",
            "RbcWKky8v8Ab4RE8JTfgaxpOvwjFuoONmOr180xUkJ8=",
            "Xh8xHgk7L7Lnf1+URnAT/cAyTQXCyqqyG9FkrpYxE70=",
            "ZvFuIEElV82sk28P5wpoUtdA+hVgjBH+eA7sv+Smleo=",
            "GrfYwQ1pBzT3PKZypdmqsTt/iishvWPfODuqYit5BrA=",
            "XQaaJ2db4DZ61rKlFckQ8Ebh9TDItVz2fX6aABPCIOA=",
            "JrZU75bHC8kO4u24orKA936FvCUUKrQ9nX4wyGyf2oE=",
            "Yf0KNRQvpU5eUazOp61yyTzte4kp/AkUx5uokleLp1c=",
            "mYkFdqU5C9m3w88wTn9l6UpYo+PI93c232s+vBiBEpM=",
            "ygTG08vB2EEmJQPXRvF5H06Tlac72AkODENK7PhCPkw=",
            "xH/DCseLHr9eKoYT8qsORZK7zVdEGYWHuVtsVrD95wY="
          ],
          "checkpoint": {
            "envelope": "rekor.sigstore.dev - 1193050959916656506\n2331802244\nFoD8lX0hTW8FS7Qv5jtr/GnckUxCUSFL9JadwU76yPM=\n\n— rekor.sigstore.dev wNI9ajBEAiAE4jyCh1mro++3Ts7h45fUxExIibPCtjADlr/13sZCVgIgR+c3+7zphnqQaK1oit8mSqcDCofoeRyX2joAdwi8X0Q=\n"
          }
        },
        "canonicalizedBody": "eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoiaGFzaGVkcmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiJiN2FkYzg0OTQwNDAxNzJjZGQwNTdmYzFjYzczMTI2ZmFiMGFlMjQyMWI0MDE0Y2JkZjRkYTBiYTgxNzZkMDE3In19LCJzaWduYXR1cmUiOnsiY29udGVudCI6Ik1FUUNJQWgzTHZ3d3pxWmFhTDNGNmN3WmE1dnFGUVcwV0hhRkh3cUwwV1R0NEl4OEFpQTRTWmZ4bm9WWURPeHRFdGdGRElMRXhqdmFrL3lOd2tvSWVJVFVxMURib1E9PSIsInB1YmxpY0tleSI6eyJjb250ZW50IjoiTFMwdExTMUNSVWRKVGlCUVZVSk1TVU1nUzBWWkxTMHRMUzBLVFVacmQwVjNXVWhMYjFwSmVtb3dRMEZSV1VsTGIxcEplbW93UkVGUlkwUlJaMEZGZFM5R1FtSndlak5QS3pVcmVIZ3ZMelF2WVdGamFGbzJkbFZrWmdwWVNHWm5NMnRoWVcxdWRsRndVV2g2TmtaUVVGWnVjRzE2ZW5ORk56WmpXV1JaUmt4TlJuaGtOVUYwVWtJclRqWnRlVko1VlcxUVNuVlJQVDBLTFMwdExTMUZUa1FnVUZWQ1RFbERJRXRGV1MwdExTMHRDZz09In19fX0="
      }
    ],
    "timestampVerificationData": {
      "rfc3161Timestamps": [
        {
          "signedTimestamp": "MIICyTADAgEAMIICwAYJKoZIhvcNAQcCoIICsTCCAq0CAQMxDTALBglghkgBZQMEAgEwgbgGCyqGSIb3DQEJEAEEoIGoBIGlMIGiAgEBBgkrBgEEAYO/MAIwMTANBglghkgBZQMEAgEFAAQgcSTKfqGLAk5upYqaptWIgSSd+kdR75L5jL/wLW5HsVgCFQDW9tQEt8zXHYsMUmyi9AwJJSAP0xgPMjAyNjA4MTMxNzI0NTJaMAMCAQGgMqQwMC4xFTATBgNVBAoTDHNpZ3N0b3JlLmRldjEVMBMGA1UEAxMMc2lnc3RvcmUtdHNhoAAxggHaMIIB1gIBATBRMDkxFTATBgNVBAoTDHNpZ3N0b3JlLmRldjEgMB4GA1UEAxMXc2lnc3RvcmUtdHNhLXNlbGZzaWduZWQCFDoTVC8MkGHuvMFDL8uKjosqI4sMMAsGCWCGSAFlAwQCAaCB/DAaBgkqhkiG9w0BCQMxDQYLKoZIhvcNAQkQAQQwHAYJKoZIhvcNAQkFMQ8XDTI2MDgxMzE3MjQ1MlowLwYJKoZIhvcNAQkEMSIEIPU/UsKPOOAIa51Xwh9TrLPfU+cDDANCHlujWSmMRCVoMIGOBgsqhkiG9w0BCRACLzF/MH0wezB5BCCF+Se8B6tiysO0Q1bBDvyBssaIP9p6uebYcNnROs0FtzBVMD2kOzA5MRUwEwYDVQQKEwxzaWdzdG9yZS5kZXYxIDAeBgNVBAMTF3NpZ3N0b3JlLXRzYS1zZWxmc2lnbmVkAhQ6E1QvDJBh7rzBQy/Lio6LKiOLDDAKBggqhkjOPQQDAgRmMGQCMDbay+j9T0zLpwG9hplCMaD6bqyIU/qsK3yer/f5fJ09/aMj4ywewGp+N+BW73VkjwIwUkbzja8LEyHwjoS0YLGp1Ep9jpOzArnq2N73gIti8ns/ZEfjYi0psruK4QxFrggi"
        }
      ]
    }
  }
}
```

**Comment by tuminoid** [comment]
> BYOK will work as well, the different bundle format and signing codepath didn't remove any features. You'll see in the bundle a "key hint" along with the signature. Here's an example, and if you wanted entirely offline signing, you could do so as well where you wouldn't fetch a timestamp or log the signature.

I dug into it, and our completely external signer (= they don't run cosign at all) can work still, if I build the bundle myself and use oras or similar to attach the bundle, then cosign verify would work. Though it seems Kyverno nor Cri-o etc don't support bundles yet, so its just theoretic. (Sigstore policy controller probably supports it, but...)

Without importing keys into cosign, and still wanting to use cosign, I could build a kms signing plugin and do: `cosign sign --key://myplugin://leaf ...` .

This is getting badly off-topic from the original issue, just wanted to conclude the thread. If I'm way off here, lets continue in Slack.
